src/tz/system/android.rs

```diff
@@ -0,0 +1,309 @@
+                err = escape::Bytes(&dlerror_message()),
+            );
+            return None;
+        };
+
+        // SAFETY: Our `SystemPropertyFind` type definition matches what is
+        // declared in `include/sys/system_properties.h` on Android.
+        let system_property_find: SystemPropertyFind =
+            unsafe { load_symbol(libc, cstr("__system_property_find\0"))? };
+
+        // SAFETY: Our `SystemPropertyRead` type definition matches what is
+        // declared in `include/sys/system_properties.h` on Android.
+        let system_property_read: SystemPropertyRead = unsafe {
+            load_symbol(libc, cstr("__system_property_read_callback\0"))?
+        };
+
+        Some(PropertyGetter {
+            libc,
+            system_property_find,
+            system_property_read,
+        })
+    }
+
+    /// Reads the given property name into the `Vec<u8>` returned.
+    ///
+    /// If the property doesn't exist, then `None` is returned and a WARN-level
+    /// log message is emitted explaining why.
+    fn get(&self, name: &CStr) -> Option<Vec<u8>> {
+        unsafe extern "C" fn callback(
+            buf: *mut c_void,
+            _name: *const i8,
+            value: *const i8,
+            _serial: u32,
+        ) {
+            let buf = buf.cast::<Vec<u8>>();
+            // SAFETY: The implied contract of `__system_property_read_callback`
+            // is that `value` is a valid NUL terminated C string.
+            let value = unsafe { CStr::from_ptr(value) };
+            // SAFETY: We passed a valid `*mut Vec<u8>` to the callback, so
+            // casting it back to it is safe.
+            unsafe {
+                (*buf).extend_from_slice(value.to_bytes());
+            }
+        }
+
+        // SAFETY: `name` is a valid NUL terminated string and
+        // `system_property_find` is a valid function read from `dlsym`
+        // according to the declaration in `include/sys/system_properties.h`.
+        let prop_info = unsafe { (self.system_property_find)(name.as_ptr()) };
+        if prop_info.is_null() {
+            warn!(
+                "Android property name `{name}` not found",
+                name = escape::Bytes(name.to_bytes()),
+            );
+            return None;
+        }
+
+        // N.B. A `prop_info` is an opaque pointer[1]... which means the
+        // implementation is probably allocating something in order to create
+        // it... right? But there's no API to free the pointer returned. And
+        // no other indication as to its lifetime. Once again, C is awful.
+        //
+        // [1]: https://android.googlesource.com/platform/bionic/+/master/libc/include/sys/system_properties.h#44
+
+        let mut buf = Vec::new();
+        // SAFETY: `name` is a valid NUL terminated string and
+        // `system_property_find` is a valid function read from `dlsym`
+        // according to the declaration in `include/sys/system_properties.h`.
```

Comment:
I believe this is meant to refer to `system_property_read` & `prop_info` being valid instead?